# Neural Network Sample
TensorFlow と Keras による IRIS の分類

## 準備
ライブラリとデータの読み込み

In [ ]:
import tensorflow as tf

import keras
from keras.layers import Dense, Activation

from sklearn import datasets
iris = datasets.load_iris()

In [ ]:
iris.data

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [ ]:
iris.target

## データの分離
学習用 80% 　テスト用 20%

In [ ]:
from sklearn.model_selection import train_test_split as split
X_train, X_test, y_train, y_test = split(iris.data, iris.target, test_size=0.2)

## モデルの作成
入力層 4：sepal_length 	sepal_width 	petal_length 	petal_width  
隠れ層 32  
出力層 3：setosa, versicolor,  virginica

In [ ]:
model = keras.models.Sequential()
model.add(Dense(units=32, input_shape=(4,), activation = 'relu'))
model.add(Dense(units=3, activation = 'softmax'))
model.compile(optimizer='sgd', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

##機械学習
epocs は学習回数で、ここでは120個（全体150個の80%）の  
データの学習を100回繰り返し行っています。

In [ ]:
model.fit(X_train, y_train, epochs=100)

Epoch 1/100
120/120 [==============================] - 0s 2ms/step - loss: 1.0028 - acc: 0.5750
Epoch 2/100
120/120 [==============================] - 0s 182us/step - loss: 0.8790 - acc: 0.6667
Epoch 3/100
120/120 [==============================] - 0s 162us/step - loss: 0.8457 - acc: 0.6917
Epoch 4/100
120/120 [==============================] - 0s 183us/step - loss: 0.8195 - acc: 0.7750
Epoch 5/100
120/120 [==============================] - 0s 167us/step - loss: 0.7984 - acc: 0.7250
Epoch 6/100
120/120 [==============================] - 0s 163us/step - loss: 0.7781 - acc: 0.6667
Epoch 7/100
120/120 [==============================] - 0s 173us/step - loss: 0.7562 - acc: 0.7083
Epoch 8/100
120/120 [==============================] - 0s 164us/step - loss: 0.7373 - acc: 0.6750
Epoch 9/100
120/120 [==============================] - 0s 167us/step - loss: 0.7211 - acc: 0.7333
Epoch 10/100
120/120 [==============================] - 0s 140us/step - loss: 0.7055 - acc: 0.7333
Epoch 11/100
120/120 

## テスト用のデータでモデルの評価
以下のとおり、テストデータでは96.7%の正解率になることがわかります

In [ ]:
score = model.evaluate(X_test, y_test, batch_size = 1)
print("正解率(accuracy)=", score[1])

30/30 [==============================] - 0s 4ms/step
正解率(accuracy)= 0.9666666666666667


##任意のデータを分類させる
事例として 4個の入力を、6.4,  2.7,  5.3,  1.9 として、正しく分類するかを試します。  
ちなみにこれは virginica 種のデータです。

In [ ]:
import numpy as np
x = np.array([[6.4, 2.7, 5.3, 1.9]])
r = model.predict(x)

r

array([[0.00543651, 0.2909139 , 0.7036496 ]], dtype=float32)

## 結果
出力された３つの数字はそれぞれ、  
  
setosa種であるの確率：0.5%  
versicolor種である確率：29.1%   
virginica種である確率：70.4%
  
を意味します。  
virginia種であることを正しく言い当てています。